### PR TITLE
HLRC support for Index Templates V2

### PR DESCRIPTION
**Pull request by probakowski**
This change adds High Level Rest Client support for Index Templates V2.

Relates to #53101

Added >non-issue label as component templates are still behind feature flag

### CODE REVIEW

File: client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java [hunk 1]
```diff
@@ -909,6 +914,36 @@ public Cancellable putTemplateAsync(PutIndexTemplateRequest putIndexTemplateRequ
...
+     * Asynchronously puts an index template using the Index Templates API.
+     * See <a href="https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-templates.html"> Index Templates API
+     * on elastic.co</a>
+     * @param putIndexTemplateRequest the request
+     * @param options the request options (e.g. headers), use {@link RequestOptions#DEFAULT} if nothing needs to be customized
+     * @param listener the listener to be notified upon request completion
+     * @return cancellable that may be used to cancel the request
+     */
+    public Cancellable putIndexTemplateAsync(PutIndexTemplateV2Request putIndexTemplateRequest,
+                                        RequestOptions options, ActionListener<AcknowledgedResponse> listener) {
```
Comment (dakrone) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java:942]: Super minor indentation nit here

File: client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java [hunk 2]
```diff
@@ -1006,6 +1071,37 @@ public Cancellable existsTemplateAsync(IndexTemplatesExistRequest indexTemplates
...
+    /**
+     * Uses the Index Templates API to determine if index templates exist
+     *
+     * @param indexTemplatesRequest the request
+     * @param options the request options (e.g. headers), use {@link RequestOptions#DEFAULT} if nothing needs to be customized
+     * @return true if any index templates in the request exist, false otherwise
+     * @throws IOException in case there is a problem sending the request or parsing back the response
+     */
+    public boolean existsIndexTemplate(IndexTemplateV2ExistRequest indexTemplatesRequest,
+                                  RequestOptions options) throws IOException {
```
Comment (dakrone) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java:1083]: Another minor indentation nit here

File: client/rest-high-level/src/main/java/org/elasticsearch/client/indices/PutIndexTemplateV2Request.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    private String name;
+
+    private String cause = "";
+
+    private boolean create;
+
+    private IndexTemplateV2 indexTemplate;
+
+    /**
+     * Sets the name of the component template.
```
Comment (dakrone) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/indices/PutIndexTemplateV2Request.java:43]:
```suggestion
     * Sets the name of the index template.
```

File: client/rest-high-level/src/main/java/org/elasticsearch/client/indices/PutIndexTemplateV2Request.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    public PutIndexTemplateV2Request name(String name) {
+        if (Strings.isNullOrEmpty(name)) {
+            throw new IllegalArgumentException("name cannot be null or empty");
+        }
+        this.name = name;
+        return this;
+    }
+
+    /**
+     * The name of the component template.
```
Comment (dakrone) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/indices/PutIndexTemplateV2Request.java:54]:
```suggestion
     * The name of the index template.
```

File: client/rest-high-level/src/main/java/org/elasticsearch/client/indices/PutIndexTemplateV2Request.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+     */
+    public String name() {
+        return this.name;
+    }
+
+    /**
+     * Set to {@code true} to force only creation, not an update of an component template. If it already
```
Comment (dakrone) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/indices/PutIndexTemplateV2Request.java:61]:
```suggestion
     * Set to {@code true} to force only creation, not an update of an index template. If it already
```

File: client/rest-high-level/src/main/java/org/elasticsearch/client/indices/PutIndexTemplateV2Request.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        this.create = create;
+        return this;
+    }
+
+    public boolean create() {
+        return create;
+    }
+
+    /**
+     * The component template to create.
```
Comment (dakrone) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/indices/PutIndexTemplateV2Request.java:74]:
```suggestion
     * The index template to create.
```

File: client/rest-high-level/src/main/java/org/elasticsearch/client/indices/PutIndexTemplateV2Request.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+     */
+    public PutIndexTemplateV2Request indexTemplate(IndexTemplateV2 indexTemplate) {
+        this.indexTemplate = indexTemplate;
+        return this;
+    }
+
+    /**
+     * The cause for this component template creation.
```
Comment (dakrone) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/indices/PutIndexTemplateV2Request.java:82]:
```suggestion
     * The cause for this index template creation.
```

File: client/rest-high-level/src/test/java/org/elasticsearch/client/IndicesClientIT.java [hunk 8]
```diff
@@ -1560,4 +1568,50 @@ public void testDeleteAlias() throws IOException {
...
+            highLevelClient().indices()::deleteIndexTemplateAsync);
+        assertThat(response.isAcknowledged(), equalTo(true));
+
+        ElasticsearchStatusException statusException = expectThrows(ElasticsearchStatusException.class,
+            () -> execute(getIndexTemplatesV2Request,
+                highLevelClient().indices()::getIndexTemplate, highLevelClient().indices()::getIndexTemplateAsync));
+
+        assertThat(statusException.status(), equalTo(RestStatus.NOT_FOUND));
+
+        exist = exist = execute(indexTemplateV2ExistRequest,
```
Comment (dakrone) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/IndicesClientIT.java:1612]:
```suggestion
        exist = execute(indexTemplateV2ExistRequest,
```

File: client/rest-high-level/src/test/java/org/elasticsearch/client/indices/GetIndexTemplatesV2ResponseTests.java [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+        if (randomBoolean()) {
+            mappings = randomMappings();
+        }
+        if (randomBoolean()) {
+            aliases = randomAliases();
+        }
+        if (randomBoolean()) {
+            composedOf = Arrays.asList(generateRandomStringArray(10, 10, false, false));
+        }
+        Template template = new Template(settings, mappings, aliases);
```
Comment (dakrone) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/indices/GetIndexTemplatesV2ResponseTests.java:95]: I think these methods (the randomization) can be reused from the component template tests you already added to the HLRC?

File: client/rest-high-level/src/main/java/org/elasticsearch/client/indices/GetIndexTemplatesV2Request.java [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+package org.elasticsearch.client.indices;
+
+import org.elasticsearch.client.TimedRequest;
+import org.elasticsearch.client.Validatable;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.unit.TimeValue;
+
+/**
+ * A request to read the content of component templates
```
Comment (dakrone) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/indices/GetIndexTemplatesV2Request.java:28]:
```suggestion
 * A request to read the content of index templates
```

**Review by dakrone — APPROVED** [review]
LGTM, I left some really minor nit comments and suggestions for fixing the docs, thanks @probakowski!

**Comment by probakowski** [comment]
@elasticmachine update branch

File: client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java (hunk 1 above)
Comment (probakowski) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java]: Fixed

File: client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java [hunk 11]
```diff
@@ -1006,6 +1071,37 @@ public Cancellable existsTemplateAsync(IndexTemplatesExistRequest indexTemplates
...
+    /**
+     * Uses the Index Templates API to determine if index templates exist
+     * @param indexTemplatesExistRequest the request
+     * @param options the request options (e.g. headers), use {@link RequestOptions#DEFAULT} if nothing needs to be customized
+     * @param listener the listener to be notified upon request completion. The listener will be called with the value {@code true}
+     * @return cancellable that may be used to cancel the request
+     */
+    public Cancellable existsIndexTemplateAsync(IndexTemplateV2ExistRequest indexTemplatesExistRequest,
+                                                  RequestOptions options,
+                                                  ActionListener<Boolean> listener) {
```
[the remark this answers is not in the document]
Comment (probakowski) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java]: Fixed

File: client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java [hunk 12]
```diff
@@ -1112,6 +1208,35 @@ public Cancellable deleteTemplateAsync(DeleteIndexTemplateRequest request, Reque
...
+     * Asynchronously delete an index template using the Index Templates API
+     * See <a href="https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-templates.html"> Index Templates API
+     * on elastic.co</a>
+     * @param request  the request
+     * @param options  the request options (e.g. headers), use {@link RequestOptions#DEFAULT} if nothing needs to be customized
+     * @param listener the listener to be notified upon request completion
+     * @return cancellable that may be used to cancel the request
+     */
+    public Cancellable deleteIndexTemplateAsync(DeleteIndexTemplateV2Request request, RequestOptions options,
+                                           ActionListener<AcknowledgedResponse> listener) {
```
[the remark this answers is not in the document]
Comment (probakowski) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java]: Fixed

File: client/rest-high-level/src/main/java/org/elasticsearch/client/indices/GetIndexTemplatesV2Request.java [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+ */
+public class GetIndexTemplatesV2Request implements Validatable {
```
[the remark this answers is not in the document]
Comment (probakowski) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/indices/GetIndexTemplatesV2Request.java]: Good idea, changed

File: client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java (hunk 2 above)
Comment (probakowski) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/IndicesClient.java]: Fixed

File: client/rest-high-level/src/test/java/org/elasticsearch/client/indices/GetIndexTemplatesV2ResponseTests.java (hunk 9 above)
Comment (probakowski) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/indices/GetIndexTemplatesV2ResponseTests.java]: Yep, that's a good idea, I've changed it

**Comment by probakowski** [comment]
@elasticmachine update branch

**Pull request merged**